Doc updates for metadata cleanup and storage (#12190)
apache/druid: docs/operations/clean-metadata-store.md

  ~ specific language governing permissions and limitations
  ~ under the License.
  -->

Apache Druid relies on [metadata storage](../dependencies/metadata-storage.md) to track information on data storage, operations, and system configuration.
The metadata store includes the following:

- Segment records
- Audit records
- Supervisor records
- Rule records
- Compaction configuration records
- Datasource records created by supervisors
- Indexer task logs

When you delete some entities from Apache Druid, records related to the entity may remain in the metadata store.
If you have a high datasource churn rate, meaning you frequently create and delete many short-lived datasources or other related entities like compaction configuration or rules, the leftover records can fill your metadata store and cause performance issues.
To maintain metadata store performance, you can configure Apache Druid to automatically remove records associated with deleted entities from the metadata store.

By default, Druid automatically cleans up metadata older than 90 days.

> This applies to all metadata entities in this topic except compaction configuration records, which do not have a retention period, and indexer task logs, for which cleanup is disabled by default.

Reviewer: Might be worth pointing out the behavior for compaction configuration record here since the durationToRetain does not applies.

Reviewer: The behavior is that at every cycle of the metadata management tasks, any compaction configuration of datasources that is inactive (datasource with no used and unused segments) will be deleted.

Author: ```suggestion
This applies to all metadata entities in this topic except compaction configuration records, which do not have a retention period, and indexer task logs, for which cleanup is disabled by default.
Druid deletes compaction configuration records at every cleanup cycle for inactive datasources.

```

### Compaction configuration records

Druid retains all compaction configuration records by default, which should be suitable for most use cases.

> If you have a high datasource churn rate, that is, you create and delete short-lived datasources with high frequency, consider turning on automated cleanup of compaction configuration records.

Reviewer: ```suggestion
If you have a high datasource churn rate, that is, you create and delete short-lived datasources with high frequency, and set auto compaction configuration on those datasources then consider turning on automated cleanup of compaction configuration records.
```

> With this feature turned on, when you create a compaction configuration for some datasource before the datasource exists, for example if ingestion is ongoing, Druid may remove the compaction configuration.

Reviewer: ```suggestion
Warning: With this feature turned on, if you create a compaction configuration for some datasource before the datasource exists, for example if initial ingestion is stil ongoing, Druid may remove the compaction configuration.
```

> To prevent the configuration from being prematurely removed, wait to set `druid.coordinator.kill.compaction.on=true` until after the datasource is created.

Reviewer: ```suggestion
To prevent the configuration from being prematurely removed, wait to set auto compaction configuration on the datasource after the datasource is created.
```

Compaction configuration records in the `druid_config` table become eligible for deletion after all segments for the datasource have been killed by the kill task. Automated cleanup for compaction configuration requires a [kill task](#kill-task).

Compaction configuration cleanup uses the following configuration:
 - `druid.coordinator.kill.compaction.on`: When `true`, enables cleanup for compaction configuration records.
 - `druid.coordinator.kill.compaction.period`: Defines the frequency in [ISO 8601 format](https://en.wikipedia.org/wiki/ISO_8601#Durations) for the cleanup job to check for and delete eligible compaction configuration records. Defaults to `P1D`.

> Compaction configuration records do not have a retention period. Druid deletes compaction configuration records at every cleanup cycle for inactive datasources, which do not have segments either used or unused.

Reviewer: Feel like this might fit better in the earlier paragraph. Maybe between line 119 and 120?

For more detail, see [Task logging](../configuration/index.md#task-logging).

## Disable automated metadata cleanup

Druid enables automated metadata cleanup by default. To disable this feature, set the following properties in the `coordinator/runtime.properties` file:

```

> druid.coordinator.kill.on=false

Reviewer: Might be good to add a comment beside each one indicating which table/entity is the config for

> druid.coordinator.kill.on=false
> druid.coordinator.kill.audit.on=false
> druid.coordinator.kill.supervisor.on=false
> druid.coordinator.kill.rule.on=false
> druid.coordinator.kill.compaction.on=false
> druid.coordinator.kill.datasource.on=false

Author: ```suggestion
# Keep unused segments
druid.coordinator.kill.on=false
# Keep audit records
druid.coordinator.kill.audit.on=false
# Keep supervisor records
druid.coordinator.kill.supervisor.on=false
# Keep rules records
druid.coordinator.kill.rule.on=false
# Keep compaction configuration records
druid.coordinator.kill.compaction.on=false
# Keep datasource records created by supervisors
druid.coordinator.kill.datasource.on=false
```
